sympy/series/formal.py

```diff
@@ -0,0 +1,445 @@
+
+    ``rational_algorithm`` uses :func:`apart` function for partial fraction
+    decomposition. ``apart`` by default uses 'undetermined coefficients method'.
+    By setting full=True, 'Bronstein's algorithm' can be used instead.
+
+    Looks for derivative of a function upto 4'th order(by default).
+    This can be overrided using order option.
+
+    returns a Tuple of (formula, series independent terms, order) if successful
+    otherwise None.
+
+    Examples
+    ========
+
+    >>> from sympy import log, atan, I
+    >>> from sympy.series.formal import rational_algorithm as ra
+    >>> from sympy.abc import x, k
+
+    >>> ra(1 / (1 - x), x, k)
+    (1, 0, 0)
+    >>> ra(log(1 + x), x, k)
+    (-(-1)**(-k)/k, 0, 1)
+
+    >>> ra(atan(x), x, k, full=True)
+    ((-I*(-I)**(-k)/2 + I*I**(-k)/2)/k, 0, 1)
+
+    Notes
+    =====
+
+    By setting full=True, range of admissible functions increases.
+    This option should be used carefully as it can signifcantly
+    slow down the computation as ``doit`` is performed on the :class:`RootSum`
+    object returned by the ``apart`` function. Use full=False whenever
+    possible.
+
+    See Also
+    ========
+
+    sympy.polys.partfrac.apart
+
+    References
+    ==========
+
+    .. [1] Formal Power Series - Dominik Gruntz, Wolfram Koepf
+    .. [2] Power Series in Computer Algebra - Wolfram Koepf
+    """
+
+    from sympy.polys import RootSum, apart
+
+    diff = f
+    ds = [] # list of diff
+
+    for i in range(order + 1):
+        if i:
+            diff = diff.diff(x)
+
+        if diff.is_rational_function(x):
+            coeff, sep = S.Zero, S.Zero
+
+            terms = apart(diff, x, full=full)
+            if terms.has(RootSum):
+                terms = terms.doit()
+
+            for t in Add.make_args(terms):
+                num, den = t.as_numer_denom()
+                if not den.has(x):
+                    sep += t
+                else:
+                    if isinstance(den, Mul):
+                        # m*(n*x - a)**j -> (n*x - a)**j
+                        ind = den.as_independent(x)
+                        den = ind[1]
+                        num /= ind[0]
+
+                    # (n*x - a)**j -> (x - b)
+                    den, j = den.as_base_exp()
+                    a, xterm = den.as_coeff_add(x)
+
+                    # term -> m/x**n
+                    if not a:
+                        sep += t
+                        continue
+
+                    xc = xterm[0].coeff(x)
+                    a /= -xc
+                    num /= xc**j
+
+                    ak = ((-1)**j * num * binomial(j + k - 1, k)
+                          .rewrite(factorial) / a**(j + k))
+                    coeff += ak
+
+            # Hacky, better way?
+            if coeff is S.Zero:
+                return None
+            if coeff.has(x) or coeff.has(zoo) or coeff.has(oo) or coeff.has(nan):
+                return None
+
+            from sympy.integrals import integrate
```

Comment:
Move the import out of the loop. Can this be a top-level import? If not, top of the function would be best.